### PR TITLE
Add support for pretrained encoders and transformed data

### PR DESCRIPTION
**Pull request by gclen**
First of all this project looks great! I've taken an initial stab at #12 and also tried to add support querying data that has already been transformed. If you have data that you've already transformed (e.g. a UMAP embedding), you probably don't want to rerun encoder.transform again. In this case you want to index the transformed data and query it directly.

This is just a first crack so happy to incorporate any feedback you might have!

### CODE REVIEW

**Comment by gclen** [comment]
Apologies for not commenting on the issue beforehand. I think it's a fair point about ```check_is_fitted``` and how it won't work for everything. 

I think the refit parameter should go in the ```Service()``` call. I think there should also be a parameter somewhere to avoid calling ```.transform()``` if the data has already been transformed. Do you think it is worth adding an additional parameter to ```Service()``` and keeping the ```indexed_from_transformed_data``` method?

**Comment by koaning** [comment]
I think preventing a `transfrom()` is fair, but the solution would be to have an `Identity()` transformer that just keeps the data as-is. This would also make a great default value for the encoder. I added a separate issue for this over at https://github.com/koaning/simsity/issues/14.

**Comment by koaning** [comment]
Let's add a `refit` to `Service()` then. This parameter should prevent the `.fit()` step from happening in the transformer. This would pass on some responsibility to the end-user, which means that we may want to catch some errors and proxy them in a meaningful way. Something like "Are you sure the encoder is trained?" in the error message would go a long way. 

@gclen would you want to pick this up by changing this PR? I'd also be fine with the creation of a new one.

**Comment by gclen** [comment]
I can pick this up (hopefully in the next couple days).

**Comment by gclen** [comment]
I've made the changes we talked about if you want to take another look.

File: simsity/service.py [hunk 1]
```diff
@@ -17,11 +17,15 @@ class Service:
...
+    def __init__(self, encoder, indexer, storage=None, refit=True) -> None:
         self.encoder = encoder
         self.indexer = indexer
         self.storage = storage if storage else {}
-        self._trained = False
+
+        if refit:
+            self._trained = False
+        else:
+            self._trained = True
```
Comment (koaning) [inline simsity/service.py:28]: Wouldn't this be more succinct:

```python
self._trained = not refit
```

File: simsity/service.py [hunk 2]
```diff
@@ -34,10 +38,21 @@ def train_from_dataf(self, df, features=None):
...
-        data = self.encoder.fit_transform(subset)
+        
+        if self._trained:
+            try:
+                data = self.encoder.transform(subset)
+            except Exception as e:
+                warnings.warn("Encountered error using pretrained encoder. Are you sure it is trained?")
+                raise e
+        else:
+            data = self.encoder.fit_transform(subset)
```
Comment (koaning) [inline simsity/service.py:51]: I think we're about to omit another `else` statement here to keep things succinct. 

```python
if not self._trained:
    self.encoder.fit(subset)

try:
    data = self.encoder.transform(subset)
except Exception as e:
    warnings.warn("Encountered error using pretrained encoder. Are you sure it is trained?")
    raise e
```

File: simsity/service.py [hunk 3]
```diff
@@ -50,13 +65,16 @@ def query(self, n_neighbors=10, out="list", **kwargs):
             kwargs: Arguments to pass as the query.
         """
         if not self._trained:
-            raise RuntimeError("Cannot save, Service is not trained.")
+            raise RuntimeError("Cannot query, Service is not trained.")
```
Comment (koaning) [inline simsity/service.py:68]: Nice catch 👍

**Review by koaning — CHANGES_REQUESTED** [review]
I think there's an opportunity to make the code a bit more succinct. Once those small comments are addressed I'll happily merge it in 👍

**Comment by koaning** [comment]
Once merged, this PR fill fix https://github.com/koaning/simsity/issues/12.

**Comment by gclen** [comment]
Done.

**Comment by koaning** [comment]
There seem to be merge conflicts now that the other PR went in. Could you resolve it? If so I'll merge this in and make a new release tomorrow.

**Comment by gclen** [comment]
Should be fixed now 👍

**Pull request merged**